remote-trait-object/src/port/server.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crossbeam::channel::{Receiver, Sender};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+
+pub struct Server {
+    receiver_thread: Option<thread::JoinHandle<()>>,
+    server_closed: Arc<AtomicBool>,
+}
+
+impl Server {
+    pub fn new<F>(dispatcher: F, ipc_send: Sender<String>, ipc_recv: Receiver<String>) -> Self
+    where
+        F: Fn(String) -> String + Send + 'static,
+    {
+        let server_closed = Arc::new(AtomicBool::new(false));
+        let server_closed_ = Arc::clone(&server_closed);
+        let receiver_thread = thread::Builder::new()
+            .name("port server receiver".into())
+            .spawn(move || {
+                receiver(dispatcher, ipc_send, ipc_recv, server_closed_);
+            })
+            .unwrap();
+
+        Server {
+            receiver_thread: Some(receiver_thread),
+            server_closed,
+        }
+    }
+}
+
+impl Drop for Server {
+    fn drop(&mut self) {
+        assert!(self.server_closed.load(Ordering::SeqCst));
```

Comment:
I think it might fail if the `receiver` thread is not scheduled before.

Reply:
I made shutdown waits until self.server_closed is changed.